src/DotNetWorker/Hosting/BootstrapHostBuilder.cs

```diff
@@ -0,0 +1,170 @@
+            {
+                _configureHostActions.Add(configureDelegate);
+            }
+
+            return this;
+        }
+
+        public IHostBuilder ConfigureServices(Action<HostBuilderContext, IServiceCollection> configureDelegate)
+        {
+            if (configureDelegate is null)
+            {
+                throw new ArgumentNullException(nameof(configureDelegate));
+            }
+
+            if (_defaultsComplete)
+            {
+                // After defaults are run, we want to run these immediately so that they are observable by the imperative code
+                configureDelegate(Context, _builder.Services);
+            }
+            else
+            {
+                _configureServicesActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+            }
+
+            return this;
+        }
+
+        public IHost Build()
+        {
+            // Functions configuration will never call this.
+            throw new InvalidOperationException();
+        }
+
+        public IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderContext, TContainerBuilder> configureDelegate)
+        {
+            // Functions configuration will never call this.
+            throw new InvalidOperationException();
+        }
+
+        public IHostBuilder UseServiceProviderFactory<TContainerBuilder>(IServiceProviderFactory<TContainerBuilder> factory) where TContainerBuilder : notnull
+        {
+            _serviceProviderFactory = new ServiceProviderFactoryAdapter<TContainerBuilder>(factory);
+            return this;
+        }
+
+        public IHostBuilder UseServiceProviderFactory<TContainerBuilder>(Func<HostBuilderContext, IServiceProviderFactory<TContainerBuilder>> factory) where TContainerBuilder : notnull
+        {
+            return UseServiceProviderFactory(factory(Context));
+        }
+
+        public void RunDefaultCallbacks()
+        {
+            foreach (var configureHostAction in _configureHostActions)
+            {
+                configureHostAction(_builder.Configuration);
+            }
+
+            // ConfigureAppConfiguration cannot modify the host configuration because doing so could
+            // change the environment, content root and application name which is not allowed at this stage.
+            foreach (var configureAppAction in _configureAppActions)
+            {
+                configureAppAction(Context, _builder.Configuration);
+            }
+
+            foreach (var configureServicesAction in _configureServicesActions)
+            {
+                configureServicesAction(Context, _builder.Services);
+            }
+
+            _defaultsComplete = true;
+        }
+
+        internal void ApplyServiceProviderFactory()
+        {
+            if (_serviceProviderFactory is null)
+            {
+                return;
+            }
+
+            _builder.ConfigureContainer(_serviceProviderFactory);
+        }
```

Comment:
Should we check if ` _builder` is null as well?

Reply:
Done.